### ISSUE TITLE
Question about dynamic `data_dim`

### ISSUE THREAD

**Issue by ratthachat** [Charl-AI/stochastic-caching#1]
Thanks for this great library!
In my application of modeling molecule, each data is a molecule encoding in (SMILES) string.

Each molecule always having distinc number of atoms, and one library is able to extract the 3D coordinate of each atom which I want to cache.

Therefore, for each data, the `data_dim` shape is rather dynamic of `(num_atoms, 3)` where `num_atoms` is change for each molecule, is there a way to use `stocaching` for this data?

At first, I thought of using Naive dictionary caching, but as you mentioned in the blog, this naive approach will not work for multi-workers dataloader. Another way is to use `num_workers=0` for the first epoch as mentioned here
https://discuss.pytorch.org/t/best-practice-to-cache-the-entire-dataset-during-first-epoch/19608/2

**Comment by Charl-AI** [comment]
Thanks for getting in touch. Sorry for the slow reply, there have been a lot of conference deadlines recently!

Having a dynamic `data_dim` is an interesting problem and not one that I had originally considered. Personally, I would avoid solutions that require setting `num_workers = 0` because, in my experience, it slows things down a lot. It should be possible to get some speedups using `stocaching` in this case, but it may depend on the distribution of the shapes (and how much engineering effort you are willing to put in). Here are my initial thoughts:

> pad -> cache -> crop

- Initialise the cache with `data_dims = (K, 3)`, where `K` is the maximum number of atoms of any molecule in your dataset. 

- When caching molecules, you can simply pad with zeros to change from `(num_atoms, 3)` to `(K,3)`. Then reverse this process by cropping off the padding when reading from the cache. This solution is probably the simplest to implement, but caching the padding will waste some space, especially if the distribution of `num_atoms` is long-tailed. 

- You could also reduce the average amount of wasted space by setting `K` to be less than the maximum `num_atoms` (e.g. to the 75%th percentile). This would preclude 25% of the dataset from being cached (you would have to add a line in your dataloader to check `num_atoms` and exclude molecules larger than `K` from the cache). However, the molecules that are cached would have less padding on average. You could find a good tradeoff for `K` heuristically, but I suspect it may be possible to derive a closed-form solution (or solve an optimisation problem) to find the best value :)

I hope this is helpful. I don't know too much about your field but I would be interested to know if it helps. If you derive a solution to the optimal `K` problem, let me know too!
